### PR TITLE
Fix for issue 23 + new tests

### PR DESCRIPTION
**Pull request by lemire**
Relates to https://github.com/FastFilter/xorfilter/pull/23

### CODE REVIEW

**Comment by lemire** [comment]
@mrdg Would you confirm the fix?

**Comment by lemire** [comment]
I am puzzling over the fact that this change should even be needed.

**Comment by lemire** [comment]
@mrdg  Ok. Adopting your fix.

This might be hiding a small issue in the code as the C version is nearly identical and it succeeds faster.

**Comment by RaduBerinde** [comment]
Just that there must be something wrong if for this particular size we need more than 100 iterations when the probability for an iteration to fail is supposed to be very small (and indeed I see in practice that we almost always need only one iteration).

I have been looking at `initializeParameters` and it looks very hard to parse what is going on. I simplified it a bit with the following diff and it fixed this issue (by "fixed" I mean that with this change `TestBinaryFuseNIssue23` passes even with `MaxIterations=2)`:

```diff
 func (filter *BinaryFuse[T]) initializeParameters(b *BinaryFuseBuilder, size uint32) {
- arity := uint32(3)
- filter.SegmentLength = calculateSegmentLength(arity, size)
- if filter.SegmentLength > 262144 {
-   filter.SegmentLength = 262144
- }
+ filter.SegmentLength, filter.SegmentCount = binaryFuseSegmentLengthAndCount(size)                                                                                 
  filter.SegmentLengthMask = filter.SegmentLength - 1
+ filter.SegmentCountLength = filter.SegmentCount * filter.SegmentLength
+ arrayLength := (filter.SegmentCount + 2) * filter.SegmentLength
+ filter.Fingerprints = reuseBuffer[T](&b.fingerprints, int(arrayLength))
+}
+
+func binaryFuseSegmentLengthAndCount(size uint32) (segmentLength, segmentCount uint32) {                                                                           
+ arity := uint32(3)
+ segmentLength = calculateSegmentLength(arity, size)
+ segmentLength = min(segmentLength, 262144)
+
  sizeFactor := calculateSizeFactor(arity, size)
  capacity := uint32(0)
  if size > 1 {
    capacity = uint32(math.Round(float64(size) * sizeFactor))
  }
- initSegmentCount := (capacity+filter.SegmentLength-1)/filter.SegmentLength - (arity - 1)                                                                          
- arrayLength := (initSegmentCount + arity - 1) * filter.SegmentLength
- filter.SegmentCount = (arrayLength + filter.SegmentLength - 1) / filter.SegmentLength                                                                             
- if filter.SegmentCount <= arity-1 {
-   filter.SegmentCount = 1
- } else {
-   filter.SegmentCount = filter.SegmentCount - (arity - 1)
- }
- arrayLength = (filter.SegmentCount + arity - 1) * filter.SegmentLength
- filter.SegmentCountLength = filter.SegmentCount * filter.SegmentLength
- filter.Fingerprints = reuseBuffer[T](&b.fingerprints, int(arrayLength))
+ totalSegmentCount := (capacity + segmentLength - 1) / segmentLength
+ totalSegmentCount = max(totalSegmentCount, arity)
+ segmentCount = totalSegmentCount - (arity - 1)
+ return segmentLength, totalSegmentCount
 }
```

These are the segment length transition points and corresponding lengths before:
```
size=2           segmentLength=4       segmentCount=1   
size=3           segmentLength=8       segmentCount=1   
size=8           segmentLength=8       segmentCount=1   
size=9           segmentLength=16      segmentCount=1   
size=27          segmentLength=16      segmentCount=2   
size=28          segmentLength=32      segmentCount=1   
size=91          segmentLength=32      segmentCount=3   
size=92          segmentLength=64      segmentCount=1   
size=303         segmentLength=64      segmentCount=5   
size=304         segmentLength=128     segmentCount=2   
size=1009        segmentLength=128     segmentCount=9   
size=1010        segmentLength=256     segmentCount=4   
size=3361        segmentLength=256     segmentCount=16  
size=3362        segmentLength=512     segmentCount=7   
size=11192       segmentLength=512     segmentCount=26  
size=11193       segmentLength=1024    segmentCount=12  
size=37272       segmentLength=1024    segmentCount=42  
size=37273       segmentLength=2048    segmentCount=20  
size=124117      segmentLength=2048    segmentCount=69  
size=124118      segmentLength=4096    segmentCount=34  
size=413309      segmentLength=4096    segmentCount=114 
size=413310      segmentLength=8192    segmentCount=56  
size=1376321     segmentLength=8192    segmentCount=188 
size=1376322     segmentLength=16384   segmentCount=93  
size=4583149     segmentLength=16384   segmentCount=313 
size=4583150     segmentLength=32768   segmentCount=156 
size=15261886    segmentLength=32768   segmentCount=522 
size=15261887    segmentLength=65536   segmentCount=260 
size=50822081    segmentLength=65536   segmentCount=871 
size=50822082    segmentLength=131072  segmentCount=435 
size=169237529   segmentLength=131072  segmentCount=1451
size=169237530   segmentLength=262144  segmentCount=725 
```

And these are after:
```
size=2           segmentLength=4       segmentCount=3   
size=3           segmentLength=8       segmentCount=3   
size=8           segmentLength=8       segmentCount=3   
size=9           segmentLength=16      segmentCount=3   
size=27          segmentLength=16      segmentCount=4   
size=28          segmentLength=32      segmentCount=3   
size=91          segmentLength=32      segmentCount=5   
size=92          segmentLength=64      segmentCount=3   
size=303         segmentLength=64      segmentCount=7   
size=304         segmentLength=128     segmentCount=4   
size=1009        segmentLength=128     segmentCount=11  
size=1010        segmentLength=256     segmentCount=6   
size=3361        segmentLength=256     segmentCount=18  
size=3362        segmentLength=512     segmentCount=9   
size=11192       segmentLength=512     segmentCount=28  
size=11193       segmentLength=1024    segmentCount=14  
size=37272       segmentLength=1024    segmentCount=44  
size=37273       segmentLength=2048    segmentCount=22  
size=124117      segmentLength=2048    segmentCount=71  
size=124118      segmentLength=4096    segmentCount=36  
size=413309      segmentLength=4096    segmentCount=116 
size=413310      segmentLength=8192    segmentCount=58  
size=1376321     segmentLength=8192    segmentCount=190 
size=1376322     segmentLength=16384   segmentCount=95  
size=4583149     segmentLength=16384   segmentCount=315 
size=4583150     segmentLength=32768   segmentCount=158 
size=15261886    segmentLength=32768   segmentCount=524 
size=15261887    segmentLength=65536   segmentCount=262 
size=50822081    segmentLength=65536   segmentCount=873 
size=50822082    segmentLength=131072  segmentCount=437 
size=169237529   segmentLength=131072  segmentCount=1453
size=169237530   segmentLength=262144  segmentCount=727 
```

In this case we have SegmentCount=14 instead of 12.

What is that code trying to do? If we want to have the given `capacity`, we need 14+2 segments, not 12+2.

**Comment by RaduBerinde** [comment]
37453 is another size that requires a good number of iterations, though not quite 100. This is the largest size with segment length 2048 and smallest segment count for that segment length.

**Comment by RaduBerinde** [comment]
Something like this:
```
		switch iterations % 4 {
		case 2:
			// Switch to smaller segment size.
			filter.SegmentLength /= 2
			filter.SegmentLengthMask = filter.SegmentLength - 1
			filter.SegmentCount = filter.SegmentCount*2 + 2
			filter.SegmentCountLength = filter.SegmentCount * filter.SegmentLength
		case 4:
			// Restore the calculated segment size.
			filter.SegmentLength *= 2
			filter.SegmentLengthMask = filter.SegmentLength - 1
			filter.SegmentCount = filter.SegmentCount/2 - 1
			filter.SegmentCountLength = filter.SegmentCount * filter.SegmentLength
		}
```

**Comment by RaduBerinde** [comment]
I will work on a test that tries to find "weaknesses" in the segment length formula by getting the average number of iterations for the largest `size` for each `(SegmentLength, SegmentCount)` pair.
